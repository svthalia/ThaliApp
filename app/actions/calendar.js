export const OPEN = 'CALENDAR_OPEN';
export const REFRESH = 'CALENDAR_REFRESH';
export const SUCCESS = 'CALENDAR_SUCCESS';
export const FAILURE = 'CALENDAR_FAILURE';
export const FETCHING = 'CALENDAR_FETCHING';

export function open() {
  return {
    type: OPEN,
  };
}

export function refresh(keywords) {
  return {
    type: REFRESH,
    payload: { keywords },
  };
}

export function success(eventList) {
  return {
    type: SUCCESS,
    payload: { eventList },
  };
}

export function failure() {
  return { type: FAILURE };
}

export function fetching() {
  return {
    type: FETCHING,
  };
}
