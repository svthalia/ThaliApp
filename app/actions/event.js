export const EVENT = 'EVENT_EVENT';
export const OPEN = 'EVENT_OPEN';
export const FETCHING = 'EVENT_FETCHING';
export const SUCCESS = 'EVENT_SUCCESS';
export const DONE = 'EVENT_DONE';
export const FAILURE = 'EVENT_FAILURE';
export const ADMIN = 'EVENT_ADMIN';
export const UPDATE_REGISTRATION = 'EVENT_UPDATE_REGISTRATION';
export const ADD_TO_CALENDAR = 'EVENT_ADD_CALENDAR';

export function event(pk, navigateToEventScreen = true) {
  return {
    type: EVENT,
    payload: { pk, navigateToEventScreen },
  };
}

export function open() {
  return {
    type: OPEN,
  };
}

export function success(eventData, eventRegistrations) {
  return {
    type: SUCCESS,
    payload: { eventData, eventRegistrations },
  };
}

export function done() {
  return {
    type: DONE,
  };
}

export function fetching() {
  return {
    type: FETCHING,
  };
}

export function failure() {
  return {
    type: FAILURE,
  };
}

export function admin() {
  return {
    type: ADMIN,
  };
}

export function updateRegistration(pk, present, payment) {
  return {
    type: UPDATE_REGISTRATION,
    payload: { pk, present, payment },
  };
}

export function addToCalendar(eventName, eventLocation, eventStart, eventEnd) {
  return {
    type: ADD_TO_CALENDAR,
    payload: {
      eventName,
      eventLocation,
      eventStart,
      eventEnd,
    },
  };
}
