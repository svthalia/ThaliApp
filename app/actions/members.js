export const OPEN = 'MEMBERS_OPEN';
export const MEMBERS = 'MEMBERS_MEMBERS';
export const FETCHING = 'MEMBERS_FETCHING';
export const MEMBERS_SUCCESS = 'MEMBERS_MEMBERS_SUCCESS';
export const FAILURE = 'MEMBERS_FAILURE';
export const MORE = 'MEMBERS_MORE';
export const MORE_SUCCESS = 'MEMBERS_MORE_SUCCESS';

export function open() {
  return {
    type: OPEN,
  };
}

export function members(keywords = '') {
  return {
    type: MEMBERS,
    payload: { keywords },
  };
}

export function fetching() {
  return {
    type: FETCHING,
  };
}

export function success(memberList, next, searchKey) {
  return {
    type: MEMBERS_SUCCESS,
    payload: { memberList, next, searchKey },
  };
}

export function failure() {
  return {
    type: FAILURE,
  };
}

export function more(url) {
  return {
    type: MORE,
    payload: { url },
  };
}

export function moreSuccess(memberList, next) {
  return {
    type: MORE_SUCCESS,
    payload: { memberList, next },
  };
}
