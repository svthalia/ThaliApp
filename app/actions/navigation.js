export const BACK = 'NAVIGATE_BACK';
export const TOGGLE_DRAWER = 'NAVIGATE_TOGGLE_DRAWER';
export const OPEN_WEBSITE = 'NAVIGATE_OPEN_WEBSITE';

export function goBack() {
  return {
    type: BACK,
  };
}

export function toggleDrawer() {
  return {
    type: TOGGLE_DRAWER,
  };
}

export function openWebsite(url) {
  return {
    type: OPEN_WEBSITE,
    payload: url,
  };
}
