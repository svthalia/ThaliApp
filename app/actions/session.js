export const INIT = 'SESSION_INIT';
export const SIGN_IN = 'SESSION_SIGN_IN';
export const SIGNED_IN = 'SESSION_SIGNED_IN';
export const SIGN_OUT = 'SESSION_SIGN_OUT';
export const TOKEN_INVALID = 'SESSION_TOKEN_INVALID';
export const FETCH_USER_INFO = 'SESSION_FETCH_USER_INFO';
export const SET_USER_INFO = 'SESSION_SET_USER_INFO';

export function init() {
  return { type: INIT };
}

export function signIn() {
  return { type: SIGN_IN };
}

export function signOut() {
  return { type: SIGN_OUT };
}

export function signedIn(accessToken, refreshToken, tokenExpiration) {
  return { type: SIGNED_IN, payload: { accessToken, refreshToken, tokenExpiration } };
}

export function tokenInvalid() {
  return { type: TOKEN_INVALID };
}

export function fetchUserInfo() {
  return { type: FETCH_USER_INFO };
}

export function setUserInfo(pk, displayName, photo) {
  return { type: SET_USER_INFO, payload: { pk, displayName, photo } };
}
