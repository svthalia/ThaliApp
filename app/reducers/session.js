import { DEFAULT_PROFILE_PHOTO } from '../constants';

import * as sessionActions from '../actions/session';

export const STATUS_SIGNED_OUT = 'SIGNED_OUT';
export const STATUS_SIGNED_IN = 'SIGNED_IN';
export const STATUS_SIGNING_IN = 'SIGNING_IN';

const initialState = {
  status: STATUS_SIGNED_OUT,
  token: '',
  displayName: '',
  pk: -1,
  photo: DEFAULT_PROFILE_PHOTO,
};

export default function session(state = initialState, action = {}) {
  switch (action.type) {
    case sessionActions.SIGN_IN:
      return {
        ...state,
        status: STATUS_SIGNING_IN,
      };
    case sessionActions.SIGNED_IN:
      return {
        ...state,
        status: STATUS_SIGNED_IN,
        ...action.payload,
      };
    case sessionActions.SET_USER_INFO:
      return {
        ...state,
        pk: action.payload.pk,
        displayName: action.payload.displayName,
        photo: action.payload.photo,
      };
    case sessionActions.TOKEN_INVALID:
    case sessionActions.SIGN_OUT:
      return initialState;
    default:
      return state;
  }
}
