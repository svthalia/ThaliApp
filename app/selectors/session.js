import { STATUS_SIGNED_IN } from '../reducers/session';

export const tokenSelector = (state) => state.session.token;
export const loggedInSelector = (state) => state.session.status === STATUS_SIGNED_IN;
