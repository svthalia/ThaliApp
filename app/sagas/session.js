import {
  call, put, takeEvery, select, delay,
} from 'redux-saga/effects';
import AsyncStorage from '@react-native-community/async-storage';
import Snackbar from 'react-native-snackbar';
import * as Sentry from '@sentry/react-native';
import i18next from '../utils/i18n';

import { apiRequest } from '../utils/url';
import * as sessionActions from '../actions/session';
import * as pushNotificationsActions from '../actions/pushNotifications';
import { tokenSelector } from '../selectors/session';
import reportError from '../utils/errorReporting';

export const IDENTIFIERKEY = '@MyStore:identifier';
export const USERNAMEKEY = '@MyStore:username';
export const TOKENKEY = '@MyStore:token';
export const DISPLAYNAMEKEY = '@MyStore:displayName';
export const PHOTOKEY = '@MyStore:photo';
export const PUSHCATEGORYKEY = '@MyStore:pushCategories';

const pairsToObject = (obj, pair) => {
  const obj2 = { ...obj };
  obj2[pair[0]] = pair[1];
  return obj2;
};

const t = i18next.getFixedT(undefined, 'sagas/session');

function* init() {
  try {
    const result = yield call(
      [AsyncStorage, 'multiGet'],
      [
        IDENTIFIERKEY,
        USERNAMEKEY,
        TOKENKEY,
        DISPLAYNAMEKEY,
        PHOTOKEY,
        PUSHCATEGORYKEY,
      ],
    );
    const values = result.reduce(pairsToObject, {});

    const id = parseInt(values[IDENTIFIERKEY], 10);
    const username = values[USERNAMEKEY];
    const token = values[TOKENKEY];
    const displayName = values[DISPLAYNAMEKEY];
    const photo = values[PHOTOKEY];
    const pushCategories = JSON.parse(values[PUSHCATEGORYKEY]);

    if (username !== null && token !== null) {
      yield put(sessionActions.signedIn(username, token));
      yield put(sessionActions.setUserInfo(id, displayName, photo));
      yield put(sessionActions.fetchUserInfo());
      yield put(pushNotificationsActions.register(pushCategories));
    } else {
      yield put(sessionActions.tokenInvalid());
    }
  } catch (e) {
    yield call(reportError, e);
  }
}

function* signIn(action) {
  const { user, pass } = action.payload;
  const data = {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      username: user,
      password: pass,
    }),
  };

  const currentTimestamp = Date.now();
  try {
    const response = yield call(apiRequest, 'token-auth', data);
    const { token } = response;

    yield call(AsyncStorage.multiSet, [
      [USERNAMEKEY, user],
      [TOKENKEY, token],
    ]);
    yield put(sessionActions.signedIn(user, token));
    yield put(sessionActions.fetchUserInfo());
    yield put(pushNotificationsActions.register());
    yield call([Snackbar, 'show'], { text: t('Login successful') });
  } catch (e) {
    // Delay failure to make sure animation is finished
    const now = Date.now();
    if (now - currentTimestamp < 150) {
      yield delay(now - currentTimestamp);
    }

    yield put(sessionActions.tokenInvalid());
    yield call(reportError, e);
    yield call([Snackbar, 'show'], { text: t('Login failed') });
  }
}

function* clearUserInfo() {
  yield call(
    [AsyncStorage, 'multiRemove'],
    [
      IDENTIFIERKEY,
      USERNAMEKEY,
      TOKENKEY,
      DISPLAYNAMEKEY,
      PHOTOKEY,
      PUSHCATEGORYKEY,
    ],
  );
  yield put(pushNotificationsActions.invalidate());
}

function* signOut() {
  yield call(clearUserInfo);
  yield call([Snackbar, 'show'], { text: t('Logout successful') });
}

function* signedIn({ payload }) {
  const { username } = payload;
  yield call(Sentry.setContext, 'user', { username });
}

function* userInfo() {
  const token = yield select(tokenSelector);

  const data = {
    method: 'GET',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      Authorization: `Token ${token}`,
    },
  };

  try {
    const userProfile = yield call(apiRequest, 'members/me', data);

    yield call(AsyncStorage.multiSet, [
      [IDENTIFIERKEY, userProfile.pk.toString()],
      [DISPLAYNAMEKEY, userProfile.display_name],
      [PHOTOKEY, userProfile.avatar.medium],
    ]);
    yield put(
      sessionActions.setUserInfo(
        userProfile.pk,
        userProfile.display_name,
        userProfile.avatar.medium,
      ),
    );
  } catch (error) {
    yield call(reportError, error);
  }
}

export default function* () {
  yield takeEvery(sessionActions.INIT, init);
  yield takeEvery(sessionActions.SIGN_IN, signIn);
  yield takeEvery(sessionActions.SIGN_OUT, signOut);
  yield takeEvery(sessionActions.SIGNED_IN, signedIn);
  yield takeEvery(sessionActions.FETCH_USER_INFO, userInfo);
  yield takeEvery(sessionActions.TOKEN_INVALID, clearUserInfo);
}
