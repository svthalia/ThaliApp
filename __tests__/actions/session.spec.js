import * as actions from '../../app/actions/session';

describe('session actions', () => {
  it('should expose the session actions', () => {
    expect(actions.INIT).toEqual('SESSION_INIT');
    expect(actions.SIGNED_IN).toEqual('SESSION_SIGNED_IN');
    expect(actions.SIGN_IN).toEqual('SESSION_SIGN_IN');
    expect(actions.TOKEN_INVALID).toEqual('SESSION_TOKEN_INVALID');
    expect(actions.SIGN_OUT).toEqual('SESSION_SIGN_OUT');
    expect(actions.FETCH_USER_INFO).toEqual('SESSION_FETCH_USER_INFO');
    expect(actions.SET_USER_INFO).toEqual('SESSION_SET_USER_INFO');
  });

  it('should create an action to init the session', () => {
    expect(actions.init()).toMatchSnapshot();
  });

  it('should create an action to notify invalid token', () => {
    expect(actions.tokenInvalid()).toMatchSnapshot();
  });

  it('should create an action to log the user in', () => {
    expect(actions.signIn()).toMatchSnapshot();
  });

  it('should create an action for a successful login', () => {
    expect(actions.signedIn('accessToken', 'refreshToken')).toMatchSnapshot();
  });

  it('should create an action to log the user out', () => {
    expect(actions.signOut()).toMatchSnapshot();
  });

  it('should create an action to fetch user info', () => {
    expect(actions.fetchUserInfo()).toMatchSnapshot();
  });

  it('should create an action to set user info', () => {
    expect(actions.setUserInfo('pk', 'displayName', 'photo')).toMatchSnapshot();
  });
});
