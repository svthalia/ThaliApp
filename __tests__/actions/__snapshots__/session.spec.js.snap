// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`session actions should create an action for a successful login 1`] = `
Object {
  "payload": Object {
    "accessToken": "accessToken",
    "refreshToken": "refreshToken",
    "tokenExpiration": undefined,
  },
  "type": "SESSION_SIGNED_IN",
}
`;

exports[`session actions should create an action to fetch user info 1`] = `
Object {
  "type": "SESSION_FETCH_USER_INFO",
}
`;

exports[`session actions should create an action to init the session 1`] = `
Object {
  "type": "SESSION_INIT",
}
`;

exports[`session actions should create an action to log the user in 1`] = `
Object {
  "type": "SESSION_SIGN_IN",
}
`;

exports[`session actions should create an action to log the user out 1`] = `
Object {
  "type": "SESSION_SIGN_OUT",
}
`;

exports[`session actions should create an action to notify invalid token 1`] = `
Object {
  "type": "SESSION_TOKEN_INVALID",
}
`;

exports[`session actions should create an action to set user info 1`] = `
Object {
  "payload": Object {
    "displayName": "displayName",
    "photo": "photo",
    "pk": "pk",
  },
  "type": "SESSION_SET_USER_INFO",
}
`;
